lib/maze/hooks/browser_hooks.rb

```diff
@@ -4,105 +4,24 @@ module Hooks
-          tunnel_id
-        when :bs
-          # BrowserStack browser
-          tunnel_id = SecureRandom.uuid
-          browsers = YAML.safe_load(File.read("#{__dir__}/../client/selenium/bs_browsers.yml"))
-          browser = browsers[config.browser]
-
-          if Maze.config.legacy_driver?
-            capabilities = Selenium::WebDriver::Remote::Capabilities.new
-            capabilities['browserstack.local'] = 'true'
-            capabilities['browserstack.localIdentifier'] = tunnel_id
-            capabilities['browserstack.console'] = 'errors'
-
-            # Convert W3S capabilities to JSON-WP
-            capabilities['browser'] = browser['browserName']
-            capabilities['browser_version'] = browser['browserVersion']
-            capabilities['device'] = browser['device']
-            capabilities['os'] = browser['os']
-            capabilities['os_version'] = browser['osVersion']
-
-            capabilities.merge! JSON.parse(config.capabilities_option)
-            config.capabilities = capabilities
-          else
-            capabilities = {
-              'bstack:options' => {
-                'local' => 'true',
-                'localIdentifier' => tunnel_id
-              }
-            }
-            capabilities.deep_merge! browser
-            capabilities.deep_merge! JSON.parse(config.capabilities_option)
-            config.capabilities = Selenium::WebDriver::Remote::Capabilities.new capabilities
-          end
-
-          Maze::Client::BrowserStackClientUtils.start_local_tunnel config.bs_local,
-                                                                   tunnel_id,
-                                                                   config.access_key
-        end
+        session_uuid = SecureRandom.uuid
 
-        # Create and start the relevant driver
-        case config.farm
+        case Maze.config.farm
         when :bb
-          # TODO: This probably needs to be settable via an environment variable
-          # selenium_url = 'https://us-west-desktop-hub.bitbar.com/wd/hub'
-          selenium_url = 'https://eu-desktop-hub.bitbar.com/wd/hub'
-          Maze.driver = Maze::Driver::Browser.new :remote, selenium_url, config.capabilities
+          @client = Maze::Client::Selenium::BitBarClient.new session_uuid
         when :bs
-          selenium_url = "http://#{config.username}:#{config.access_key}@hub.browserstack.com/wd/hub"
-          Maze.driver = Maze::Driver::Browser.new :remote, selenium_url, config.capabilities
+          @client = Maze::Client::Selenium::BrowserStackClient.new session_uuid
         when :local
-          Maze.driver = Maze::Driver::Browser.new Maze.config.browser.to_sym
+          @client = Maze::Client::Selenium::LocalClient.new session_uuid
         end
 
-        # Write links to device farm sessions, where applicable
-        write_session_link
+        @client.prepare_session
+        @client.start_session
+        @client.log_session_info
```

Comment:
If this is the only place using the clients, does it make sense to have three separate methods? Could the others be put into `start_session` so this can be simpler:

```suggestion
        @client.start_session
```

Then the client interface can be simplified to match:

```ruby
class Client
  def start_session -> void
  def end_session -> void
end
```

Reply:
Yes, you're right.  I following the implementation for the Appium too closely (which was only done to that granularity because the code bounces around between command and farm-specifics).